[dtensor] refactor shape/offset calculation

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95924
* __->__ #95923

Shape offset calculation is commonly used and extract them into a separate util

wanchaol committed:
[dtensor] refactor shape/offset calculation

[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/_utils.py:
```diff
@@ -0,0 +1,66 @@
+                my_coordinate[idx],
+                return_offset=True,
+            )
+            local_shape[shard_dim] = shard_size
```
Is this line needed?

wanchaol replied:
Yep it's needed because we need to use the updated local_shape to calculate the next offset

fduwjj commented on torch/distributed/_tensor/_utils.py:
```diff
@@ -0,0 +1,66 @@
+        mesh_dim_size = mesh.size(idx)
+        my_coordinate = mesh.get_coordinate()
+        assert my_coordinate is not None, "Rank not part of mesh!"
+        if isinstance(placement, Shard):
+            shard_dim = placement.dim
+            assert (
+                shard_dim < ndim
+            ), f"Sharding dim {shard_dim} greater than tensor ndim {ndim}"
+            local_shard_size, _ = placement._local_shard_size_on_dim(
+                local_shape[shard_dim], mesh_dim_size, my_coordinate[idx]
+            )
```
Nit: might further consolidate the duplicate logic?

wanchaol replied:
Yeah we can see how to consolidate the logic further, I was thinking if I should merge these two APIs, but since they are for different purposes, we can separate them for now.

fduwjj reviewed (approved):
LGTM

wanchaol committed:
Update on "[dtensor] refactor shape/offset calculation"


Shape offset calculation is commonly used and extract them into a separate util



[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] refactor shape/offset calculation"


Shape offset calculation is commonly used and extract them into a separate util



[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] refactor shape/offset calculation"


Shape offset calculation is commonly used and extract them into a separate util



[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
